File: distros/J/JavonetPerlSdk/lib/Javonet/Binaries/Python/javonet/core/protocol/CommandSerializer.py
from javonet.core.protocol.TypeSerializer import TypeSerializer
from javonet.utils.Command import Command
from javonet.utils.RuntimeName import RuntimeName
from javonet.utils.connectionData.IConnectionData import IConnectionData


class CommandSerializer:
    buffer = []

    def serialize(self, root_command: Command, connection_data: IConnectionData, runtime_version=0):
        self.insert_into_buffer([root_command.runtime_name.value, runtime_version])
        self.insert_into_buffer(connection_data.serialize_connection_data())
        self.insert_into_buffer([RuntimeName.python.value, root_command.command_type.value])
        self.serialize_recursively(root_command)
        return self.buffer

    def serialize_recursively(self, command):
        for item in command.get_payload():
            if isinstance(item, Command):
                self.insert_into_buffer(TypeSerializer.serialize_command(item))
                self.serialize_recursively(item)
            else:
                self.insert_into_buffer(TypeSerializer.serialize_primitive(item))

        return

    def insert_into_buffer(self, arguments):
        self.buffer = self.buffer + arguments

File: distros/J/JavonetPerlSdk/lib/Javonet/Binaries/Ruby/MacOs/X64/core/protocol/command_serializer.rb
require_relative 'type_serializer'
require_relative '../../utils/command'
require_relative '../../utils/runtime_name'
require_relative '../../utils/tcp_connection_data'

class CommandSerializer
  def initialize
    @buffer = []
  end

  def serialize(root_command, connection_data = nil, runtime_version = 0)
    insert_into_buffer([root_command.runtime_name, runtime_version])
    if connection_data.nil?
      insert_into_buffer([0, 0, 0, 0, 0, 0, 0])
    else
      insert_into_buffer(connection_data.serialize_connection_data)
    end
    insert_into_buffer([RuntimeName::PYTHON, root_command.command_type])
    serialize_recursively(root_command)
    @buffer
  end

  def serialize_recursively(command)
    command.payload.each do |item|
      if item.is_a?(Command)
        insert_into_buffer(TypeSerializer.serialize_command(item))
        serialize_recursively(item)
      else
        insert_into_buffer(TypeSerializer.serialize_primitive(item))
      end
    end
  end

  def insert_into_buffer(arguments)
    @buffer += arguments
  end
end